fixes #3178 - add feature for compute profiles for hostgroup and host

isratrade opened the pull request.

ohadlevy commented on app/controllers/hardware_profile_compute_resources_controller.rb:
```diff
@@ -0,0 +1,35 @@
+                                                                            :compute_resource_id => params[:compute_resource_id])
+    @hardware_profile = HardwareProfile.find(params[:hardware_profile_id])
+    @compute_resource = ComputeResource.find(params[:compute_resource_id])
+    render :edit
```
why render edit?

isratrade replied:
@ohadlevy, I had `new` template, but it was the same as `edit`, so I didn't see the purpose of having both.  Many times, the `new` template has a different title than `edit` and both render 'form'.  In this case, the title is the same.
I can put back a `new` template to avoid confusion.  Just let me know.

ohadlevy replied:
well, you kept the new template in, imho, we should leave it.

ohadlevy commented on app/controllers/hardware_profile_compute_resources_controller.rb:
```diff
@@ -0,0 +1,35 @@
+  def new
+    @hardware_profile_compute_resource = HardwareProfileComputeResource.new(:hardware_profile_id => params[:hardware_profile_id],
+                                                                            :compute_resource_id => params[:compute_resource_id])
+    @hardware_profile = HardwareProfile.find(params[:hardware_profile_id])
```
nitpick: cant we simply access that via a the relationship? e.g. @hwp_cr.hw_pw.id ?

ohadlevy commented on app/controllers/hardware_profile_compute_resources_controller.rb:
```diff
@@ -0,0 +1,35 @@
+  end
+
+  def update
+    @hardware_profile_compute_resource = HardwareProfileComputeResource.find(params[:id])
```
nitpick: var name seems extensively too long in my taste :)

ohadlevy commented on app/controllers/hardware_profiles_controller.rb:
```diff
@@ -0,0 +1,50 @@
+  before_filter :find_hardware_profile, :only => %w{edit show update destroy}
+
+  def index
+    @hardware_profiles = HardwareProfile.paginate(:page => params[:page])
```
no search/sort?

ohadlevy commented on app/controllers/hardware_profiles_controller.rb:
```diff
@@ -0,0 +1,50 @@
+
+  private
+
+  def find_hardware_profile
```
not related, but its about time we start converging the api app controller and the ui app controller, code like find_resource can be used everywhere.

isratrade replied:
submitted issue http://projects.theforeman.org/issues/3933

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -132,7 +132,10 @@ def destroy
     return not_found unless (params[:host] && (id=params[:host][:compute_resource_id]))
     Taxonomy.as_taxonomy @organization, @location do
-      render :partial => "compute", :locals => {:compute_resource => ComputeResource.find_by_id(id)}
+      hardware_profile_id = params[:host][:hardware_profile_id] || Hostgroup.find(params[:host][:hostgroup_id]).hardware_profile_id
```
does it make more sense to add a writter/getter on the host object instead?
e.g.

``` ruby
attr_writer :hw_profile_id

def hw_profile_id
  @hw_profile_id =|| hostgroup.hw_profile_id
end
```

isratrade replied:
@ohadlevy, `hardware_profile_id` is already an attribute and column in the db on hosts.  The code above assumes that the `hardware_profile_id` is only inherited from the hostgroup and not overwritten on the host level

ohadlevy replied:
yes, I didnt see that when i wrote it, but just assume its
read_attributes(..) instead of @var

isratrade replied:
@ohadlevy, I will add this to managed.rb

``` ruby
  # take hardware_profile_id from hostgroup if blank on host
  def hardware_profile_id
    @hardware_profile_id =|| hostgroup.try(:hardware_profile_id)
  end
```

ohadlevy replied:
I dont think this will work? 
you need something like

``` ruby
def hardware_profile_id
   read_attribute(:hardware_profile_id) || hostgroup.try(:hardware_profile_id)
end
```

isratrade replied:
@ohadlevy, but I think the code on line 135 is more readable as is.  I'm inside an ajax method `compute_resource_selected` and don't have access to @host.  I need to take the params submitted by javascript on the form, `params[:host][:hardware_profile_id]` and `params[:host][:hostgroup_id]`

isratrade replied:
@ohadlevy, on 2nd thought. I removed the code in the comment above that overrides the getter `hardware_profile_id`

isratrade replied:
@ohadlevy, I put the code back as you suggested with 

``` ruby
read_attribute(:hardware_profile_id) || hostgroup.try(:hardware_profile_id)
```

ohadlevy commented on app/models/compute_resource.rb:
```diff
@@ -151,12 +153,6 @@ def vm_instance_defaults
-  def hardware_profiles(opts={})
-  end
-
-  def hardware_profile(id,opts={})
```
this might conflict with ovirt, which does have hw profiles.
afaik, this was here because sub classes can override it.

isratrade replied:
@ohadlevy, should I rename the relationship with hardware profiles to avoid this conflict?

ohadlevy commented on app/models/compute_resources/foreman/model/ec2.rb:
```diff
@@ -99,6 +99,10 @@ def associated_host(vm)
     end
 
+    def flavor_name(flavor_id)
+      "#{flavor_id} - #{flavors.find {|i| i.id == flavor_id.to_s}.try(:name)}"
```
i think you can have a more efficient query here by using fog, something like

``` ruby
client.flavors.get(flavor_id).try(:name)
```

ohadlevy commented on app/models/compute_resources/foreman/model/openstack.rb:
```diff
@@ -82,6 +82,10 @@ def associated_host(vm)
     end
 
+    def flavor_name(flavor_ref)
+      flavors.find {|i| i.id == flavor_ref.to_s}.try(:name)
```
ditto

ohadlevy commented on app/models/hardware_profile_compute_resource.rb:
```diff
@@ -0,0 +1,15 @@
+  validates :hardware_profile_id, :presence => true, :uniqueness => {:scope => :compute_resource_id}
+  validates :compute_resource_id, :presence => true, :uniqueness => {:scope => :hardware_profile_id}
+
+  serialize :vm_defaults, Hash
```
maybe just call it attributes?

ohadlevy replied:
maybe not now, but we should consider how to validate these attributes (e.g. via the api)

ohadlevy replied:
I think that if you change here from Hash to HashWithIn.. it would always be that.

ohadlevy commented on app/models/hostgroup.rb:
```diff
@@ -166,6 +174,13 @@ def nested_root_pw
   end
 
+  def nested_hardware_profile_id
+    Hostgroup.sort_by_ancestry(ancestors).reverse.each do |a|
```
maybe include eager loading, so its not n+1 queries

ohadlevy replied:
also, you can improve the query not to include other parent hg with null hw_profile_id

isratrade replied:
I changed method to

``` ruby
    Hostgroup.sort_by_ancestry(ancestors.where('hardware_profile_id > 0')).last.hardware_profile_id if ancestry.present?
```

however, I didn't see n+1 queries in the previous

ohadlevy commented on app/views/compute_resources_vms/form/_ec2.html.erb:
```diff
@@ -20,6 +20,7 @@
   if ( groups = security_groups_for_vpc(compute_resource.security_groups, vpc_id)) %>
   <%= selectable_f f, :security_group_ids, groups, {},
     { :multiple => true, :label => _("Security groups"), :disabled => !(@host.nil? || @host.ip.empty?),
+      :size => "col-md-10",
```
maybe not related, @abenari maybe this needs to be part of the ui-fixes branch?

ohadlevy commented on app/views/compute_resources_vms/form/_libvirt.html.erb:
```diff
@@ -1,5 +1,5 @@
 <% new = f.object && f.object.new? %>
-<%= text_f f, :name, :disabled => !new if controller_name != "hosts" %>
+<%= text_f f, :name, :disabled => !new if controller_name != "hosts" && controller_name != "hardware_profile_compute_resources" %>
```
maybe extract to an helper?

ohadlevy commented on app/views/hardware_profile_compute_resources/_form.html.erb:
```diff
@@ -0,0 +1,16 @@
+<%= javascript 'compute_resource', 'lookup_keys'%>
```
do we need thelookup_keys js here?

isratrade replied:
although not obvious, lookup_keys.js contains js functions `add_child_node` and `remove_child_node` that is used on forms such as _libvirt.html.erb

ohadlevy commented on db/migrate/20131224153518_create_hardware_profiles.rb:
```diff
@@ -0,0 +1,9 @@
+class CreateHardwareProfiles < ActiveRecord::Migration
+  def change
+    create_table :hardware_profiles do |t|
+      t.string :name
```
i think we need an order option, maybe at first, we should use the same prefixes like ec2 uses + long name

isratrade replied:
Should I enter seed data in the `name` field such as
1-Small
2-Medium
3-Large

Or should I create another column called `prefix` and change the default_scope to order by `prefix`

ohadlevy commented on db/migrate/20131224154836_add_hw_profile_to_hostgroup.rb:
```diff
@@ -0,0 +1,6 @@
+class AddHwProfileToHostgroup < ActiveRecord::Migration
+  def change
+    add_column :hostgroups, :hardware_profile_id, :integer
```
indexes?

isratrade replied:
added in separated migration

isratrade commented on config/routes.rb:
```diff
@@ -119,6 +119,13 @@
+  resources :hardware_profiles do
+    resources :hardware_profile_compute_resources, :path => :sets, :only => [:create, :edit, :update]
+    resources :compute_resources, :only => [] do
+      resources :hardware_profile_compute_resources, :path => :sets, :only => :new
```
only the `new` path is nested in both hardware_profiles and compute_resources.  It create a url like this.
`http://0.0.0.0:3000/hardware_profiles/1-1-small/compute_resources/8-h01-sat-lab/sets/new`

isratrade commented on db/seeds.rb:
```diff
@@ -150,3 +150,17 @@ def audit_modified?(type, name)
+    [
+      { :name => "1-Small" },
+      { :name => "2-Medium" },
+      { :name => "3-Large" },
```
@domcleal, @ohadlevy, should I also put this seeds data in the migration or not?

ohadlevy replied:
seed.

isratrade commented:
re-committed with tests added

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -132,7 +132,10 @@ def destroy
+      hardware_profile_id = params[:host][:hardware_profile_id] || Hostgroup.find(params[:host][:hostgroup_id]).hardware_profile_id
+      compute_resource = ComputeResource.find(id)
+      render :partial => "compute", :locals => {:compute_resource => compute_resource,
+                                                :hw_defaults => compute_resource.hardware_profile_defaults(hardware_profile_id) }
```
why call it defaults? i find the word attributes a better match?

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -444,6 +447,7 @@ def process_hostgroup
     @environment     = @hostgroup.environment
     @domain          = @hostgroup.domain
     @subnet          = @hostgroup.subnet
+    @hardware_profile  = @hostgroup.hardware_profile
```
nitpick: = sign identiation

ohadlevy commented on app/models/compute_resource.rb:
```diff
@@ -187,6 +191,14 @@ def set_console_password=(setpw)
   end
 
+  def hardware_profile_defaults(id)
+    HashWithIndifferentAccess.new(self.hardware_profile_compute_resources.find_by_hardware_profile_id(id).try(:vm_defaults))
```
why not serialize it in the first place as HashWithInDifferentAccess?

ohadlevy commented on app/views/compute_resources/form/_ovirt.html.erb:
```diff
@@ -9,4 +9,4 @@
                  :'data-url' => test_connection_compute_resources_path) + image_tag('/assets/spinner.gif', :id => 'test_connection_indicator', :class => 'hide').html_safe }) %>
 <% quotas = (f.object.uuid.nil? && controller.action_name != 'test_connection') ? [] : f.object.quotas rescue []%>
-<%= select_f f, :ovirt_quota, quotas, :id, :name, {}, :label => _("Quota ID") %>
+<% select_f f, :ovirt_quota, quotas, :id, :name, {}, :label => _("Quota ID") %>
```
I can see there is an error in this line, but I'm not sure what you wanted to change here?

isratrade replied:
typo.  no change here.  I commented out per Amos' help and when I uncommented, the = got accidentally deleted.

ohadlevy commented on app/views/compute_resources_vms/form/_openstack.html.erb:
```diff
@@ -1,5 +1,5 @@
 <% new = f.object && f.object.id.blank? %>
-<%= text_f f, :name, :disabled => !new if controller_name != "hosts" %>
+<%= text_f f, :name, :disabled => !new if show_vms_name? %>
```
maybe rename to show_vm_name ?

ohadlevy commented on app/views/hardware_profile_compute_resources/new.html.erb:
```diff
@@ -0,0 +1,3 @@
+<% title _("Edit hardware profile on #{@set.compute_resource.to_label}") %>
```
nitpick: title should be New ?

ohadlevy commented on app/views/hardware_profiles/show.html.erb:
```diff
@@ -0,0 +1,28 @@
+    <tr>
+      <td><%= link_to(compute_resource.to_label, which_path) %></td>
+      <td><%= pretty_vm_defaults(compute_resource.hardware_profile_compute_resources.where(:hardware_profile_id => @hardware_profile.id).first, compute_resource) %>
+          <%# debugging only - uncomment below to show full hash saved %>
```
remove ?

ohadlevy commented on app/views/hostgroups/_form.html.erb:
```diff
@@ -27,7 +27,8 @@
         {:label => _("Environment"), :onchange => 'update_puppetclasses(this);', :"data-url" => environment_selected_hostgroups_path} %>
-
+      <%= select_f(f, :hardware_profile_id, HardwareProfile.visibles, :id, :name, {:include_blank => (f.object.parent_id.present? ? _("Inherit from Parent") : '')},
+                                                                             {:label => _("Hardware Profile")}) if HardwareProfile.visibles.count > 0 %>
```
.any?

ohadlevy commented on app/views/hosts/_form.html.erb:
```diff
@@ -55,8 +55,19 @@
+            :onchange => 'computeResourceSelected(this);',
+            :help_inline => :indicator } if SETTINGS[:unattended] && @host.new_record? || @host.compute_resource_id %>
+
+        <% if @host.new_record? && HardwareProfile.visibles.count > 0 %>
```
.any?

ohadlevy commented on config/routes.rb:
```diff
@@ -251,7 +258,7 @@
       resources :compute_resources do
         member do
-          post 'hardware_profile_selected'
+          post 'template_selected'
```
don't we have JS that uses this path? I'm guessing its data-attribute but i didn't check.

isratrade replied:
@ohadlevy, it's used in /app/views/compute_resources_vms/form/_ovirt.html.erb in `data-url`

``` ruby
<%= select_f f, :template, compute_resource.templates, :id, :name, {:include_blank => _("Select template")},
             { :disabled    => !new, :'data-url' => template_selected_compute_resource_path(compute_resource),
```

ohadlevy commented on db/migrate/20131224153743_create_hardware_profile_compute_resources.rb:
```diff
@@ -0,0 +1,11 @@
+class CreateHardwareProfileComputeResources < ActiveRecord::Migration
+  def change
+    create_table :hardware_profile_compute_resources do |t|
+      t.integer :hardware_profile_id
```
indexes?

ohadlevy commented:
I think it would make sense to add auditing?

ohadlevy commented:
note, tests are failing because of non ruby 187 syntax

isratrade commented:
@ohadlevy, re-committed with changes in separate commit to squash.  I go the the following error when I tried to serialize by HashWithIndifferentAccess
`Attribute was supposed to be a ActiveSupport::HashWithIndifferentAccess, but was a Hash`
I deleted all the previous data, so that wasn't the issue.  There are several threads where is a bug 
http://stackoverflow.com/questions/8918434/unable-to-serialize-as-activesupporthashwithindifferentaccess-anymore

ohadlevy commented on app/models/hardware_profile.rb:
```diff
@@ -0,0 +1,23 @@
+class HardwareProfile < ActiveRecord::Base
+
+  include Authorization
+  audited :allow_mass_assignment => true
```
i think you can audit with associated audits, so they are tracked as one ?

isratrade replied:
`has_associated_audits` doesn't do anything since table `hardware_profiles` is not updated when join table `hardware_profile_compute_resources` is updated.

ohadlevy replied:
btw: can we not allow_mass_assignment for now? (I assume you can define the atts we allow?)

ohadlevy commented on app/models/hardware_profile_compute_resource.rb:
```diff
@@ -0,0 +1,17 @@
+  serialize :vm_attrs, Hash
+
+  def method_missing(method, *args, &block)
+    vm_attrs["#{method}"]
```
I think you should add some rescue statements here?

ohadlevy commented on app/controllers/compute_resources_controller.rb:
```diff
@@ -102,8 +102,8 @@ def test_connection
-  def hardware_profile_selected
-    compute = @compute_resource.hardware_profile(params[:hwp_id])
+  def template_selected
+    compute = @compute_resource.template(params[:hwp_id])
```
should we keep calling the param hwp_id?

are we breaking the ovirt api ? (for those who used vm attrs via the cli for example?)

isratrade replied:
@ohadlevy, I can't find where params[:hwp_id] is defined in the 'develop' branch.  It doesn't make sense that the API would use an AJAX method.

ohadlevy commented on app/helpers/audits_helper.rb:
```diff
@@ -23,6 +23,8 @@ def audit_title audit
     type_name = audited_type audit
     if type_name == "Puppet Class"
       "#{id_to_label audit.audited_changes.keys[0], audit.audited_changes.values[0]}"
+    elsif type_name == "Set (Hardware Profile / Compute Resource)"
```
nitpick: use case

ohadlevy commented on app/helpers/compute_resources_vms_helper.rb:
```diff
@@ -117,4 +117,29 @@ def compute_object_vpc_id(form)
+    when 'EC2'
+      "#{compute_resource.flavor_name(hardware_profile_compute_resource.flavor_id)}"
+    when 'Libvirt'
+      "#{hardware_profile_compute_resource.cpus} CPUs and #{number_to_human_size(hardware_profile_compute_resource.memory.to_i)} memory"
```
would this make more sense to have as to_label or something in the fog extensions mixin?

isratrade replied:
@ohadlevy, but I don't have an object Fog::Compute::AWS::Flavor to call a method on.  I'm just saving the hash of attributes in `vm_attrs` and using `vm_attrs[flavor_id]` (called by just `.flavor_id`) to get the flavor name.

ohadlevy commented on app/models/compute_resources/foreman/model/openstack.rb:
```diff
@@ -82,6 +82,10 @@ def associated_host(vm)
     end
 
+    def flavor_name(flavor_ref)
+      client.flavors.get(flavor_id).try(:name)
```
nitpick: not DRY, maybe consider having a cloud mixin ( i can see ssh keys handling and userdata also moving there)

ohadlevy commented on app/models/hardware_profile_compute_resource.rb:
```diff
@@ -0,0 +1,23 @@
+class HardwareProfileComputeResource < ActiveRecord::Base
+
+  audited :allow_mass_assignment => true
```
ditto about mass_assignment

ohadlevy commented on app/models/hardware_profile_compute_resource.rb:
```diff
@@ -0,0 +1,23 @@
+
+  def method_missing(method, *args, &block)
+    return vm_attrs["#{method}"] if vm_attrs.keys.include?(method.to_s)
+    raise "#{method} is an unknown attribute"
```
use foreman_exception ?

ohadlevy commented on app/models/hostgroup.rb:
```diff
@@ -166,6 +174,10 @@ def nested_root_pw
   end
 
+  def nested_hardware_profile_id
+    Hostgroup.sort_by_ancestry(ancestors.where('hardware_profile_id > 0')).last.try(:hardware_profile_id) if ancestry.present?
```
should we generalize this? as this repeats with nested_root_pw and would repeat in orgs? (also note the implementation is different between root_pass and hw_profile)

isratrade replied:
http://projects.theforeman.org/issues/3939

isratrade commented on app/models/hardware_profile_compute_resource.rb:
```diff
@@ -0,0 +1,51 @@
+    compute_resource.new_vm(vm_attrs)
+  end
+
+  def pretty_vm_attrs
```
model method instead of the helper method.  API/CLI could use it too.

isratrade commented:
do not merge. errors with FogExtensions::RackspaceV2

isratrade commented:
re-committed with fix to FogExtensions::RackspaceV2.
@domcleal, not sure why this integration test failed for PG???

```
HostTest.test_0003_show page
Error Message
my5name.mydomain.net <h1> tag, but was not found
```

domcleal commented:
@isratrade the integration tests only run on PG, so I'd assume it's always failing.  Try running it locally?

isratrade commented:
[test]

domcleal commented on app/models/concerns/fog_extensions/vsphere/server.rb:
```diff
@@ -23,6 +23,10 @@ def reset
       end
 
+      def vm_description
+        "#{cpus} CPUs and #{memory_mb.to_i} MB memory"
```
string should be extracted for i18n

domcleal commented on app/models/concerns/fog_extensions/ovirt/server.rb:
```diff
@@ -20,6 +22,10 @@ def reset
       end
 
+      def vm_description
+        "#{cores} Cores and #{number_to_human_size(memory)} memory"
```
extract string

domcleal commented on app/models/concerns/fog_extensions/libvirt/server.rb:
```diff
@@ -25,6 +27,10 @@ def reset
       end
 
+      def vm_description
+        "#{cpus} CPUs and #{number_to_human_size(memory)} memory"
```
extract string

domcleal commented on app/views/hardware_profile_compute_resources/edit.html.erb:
```diff
@@ -0,0 +1,3 @@
+<% title _("Edit hardware profile on #{@set.compute_resource.to_label}") %>
```
Don't interpolate inside a string like this, use %s

domcleal commented on app/views/hardware_profile_compute_resources/new.html.erb:
```diff
@@ -0,0 +1,3 @@
+<% title _("New hardware profile on #{@set.compute_resource.to_label}") %>
```
don't interpolate

domcleal commented on app/views/hardware_profiles/show.html.erb:
```diff
@@ -0,0 +1,27 @@
+<% title(_("Hardware profile") + ": " +  @hardware_profile.name) %>
```
use %s

domcleal commented on app/views/hardware_profiles/show.html.erb:
```diff
@@ -0,0 +1,27 @@
+<% title(_("Hardware profile") + ": " +  @hardware_profile.name) %>
+
+<p>
+  Click on the link of a compute resource to edit it's default vm characteristics.
```
Extract this, remove the apostrophe from "it's" and upper-case "VM" or use "host" (might be preferable).

domcleal commented on app/views/hardware_profiles/show.html.erb:
```diff
@@ -0,0 +1,27 @@
+
+<table class="table table-bordered table-striped">
+  <tr>
+    <th><%= "Compute Resource " %></th>
```
extract missing, don't include the space in the string

domcleal commented on app/views/hardware_profiles/show.html.erb:
```diff
@@ -0,0 +1,27 @@
+<table class="table table-bordered table-striped">
+  <tr>
+    <th><%= "Compute Resource " %></th>
+    <th><%= "VM Attributes (#{@hardware_profile.name})" %></th>
```
extract

domcleal commented on app/views/hostgroups/_form.html.erb:
```diff
@@ -27,7 +27,8 @@
       <%= select_f f, :environment_id, Environment.all, :id, :to_label, {:include_blank => true},
         {:label => _("Environment"), :onchange => 'update_puppetclasses(this);', :"data-url" => environment_selected_hostgroups_path} %>
-
+      <%= select_f(f, :hardware_profile_id, HardwareProfile.visibles, :id, :name, {:include_blank => (f.object.parent_id.present? ? _("Inherit from Parent") : '')},
```
Lower case "P" and "P" in Hardware profile below.

domcleal commented on app/views/hosts/_form.html.erb:
```diff
@@ -55,8 +55,20 @@
+        <div id="hardware_profile" <%= display?(!@host.compute_resource_id) %> >
+          <%= select_f f, :hardware_profile_id, HardwareProfile.visibles, :id, :name,
+            { :include_blank => true },
+            { :label => _("Hardware Profile"), :'data-url' => compute_resource_selected_hosts_path,
```
Lower case "P"?

isratrade commented:
I'm still confused by the one integration test failure.

``` ruby
    assert page.has_selector?('h1', :text => "my5name.mydomain.net"), "my5name.mydomain.net <h1> tag, but was not found"
```

isratrade commented:
@domcleal, re-committed with changes to squash before merge.

domcleal commented on app/views/hardware_profiles/show.html.erb:
```diff
@@ -0,0 +1,27 @@
+<% title(_("Hardware profile: %s") % @hardware_profile.name) %>
+
+<p>
+  Click on the link of a compute resource to edit its default VM attributes.
```
Missing string extraction

domcleal replied:
@isratrade still here?

domcleal commented:
I can reproduce the test failure quite easily here with rake test.  From some debugging, what's happening is that Fog is returning mock objects even though `Fog.mock? => false`.  I think the additional tests that rely on Fog mocking are perhaps causing side effects and Fog isn't "unmocking" entirely.

In the context of the failing test:

<pre>
[7] pry(#<HostTest>)> Fog.mock?
=> false
[8] pry(#<HostTest>)> Fog.unmock!
=> false
[9] pry(#<HostTest>)> Host.find_by_name("my5name.mydomain.net").compute_resource.send(:client).servers.get("a")
=>   Fog::Compute::Libvirt::Server
    id="dom.uuid",
etc.
</pre>

Fog is returning a server object for any parameter given to .get, so whether there's a UUID on the Host or not, we get a VM back.  So when the Host#show page renders, it's getting a `@vm` (app/views/hosts/show.html.erb:4).  This means the title of the page changes to contain "fog-dom1".

I'd suggest we embrace this and surround the integration test with Fog.mock!/unmock!, then test for the h1 containing fog-dom1 instead.

isratrade commented:
@domcleal, where is @vm in hosts#show?  I don't see it in the title method

``` ruby
<% title @host.to_label, icon(@host.os) + @host.to_label %>
```

domcleal commented:
Ah yes, I've no idea then!

domcleal commented on app/services/menu/loader.rb:
```diff
@@ -43,6 +43,7 @@ def self.load
             menu.divider                :caption => N_('Provisioning Setup')
             menu.item :operatingsystems,:caption => N_('Operating systems')
             menu.item :config_templates,:caption => N_('Provisioning templates')
+            menu.item :hardware_profiles, :caption => N_('Hardware profiles')
```
Nitpick, I'd move this near hardware models as it isn't related to prov. templates or ptables, both types of templates.

domcleal commented on app/views/hardware_profiles/show.html.erb:
```diff
@@ -0,0 +1,27 @@
+    <tr>
+      <td><%= link_to(compute_resource.to_label, which_path) %></td>
+      <td>
+        <%= compute_resource.hardware_profile_compute_resources.where(:hardware_profile_id => @hardware_profile.id).first.try(:pretty_vm_attrs) %>
```
Nitpick, maybe add some italic+grey text here saying "unspecified" or similar if no profile for this CR exists.

domcleal commented:
Some comments:
1. A help page for hardware profiles would be very useful, as would a section for the 1.4 manual (we can link from the welcome page to the manual too).  It's not clear after you create one what you need to do next, an intro may help here, possibly some info in the success notification message, and/or simply redirect from the create call to the show page for the newly created profile.
   - thanks to @ohadlevy, I found it on the "show" page.. could we make this clearer somehow?  Perhaps make an additional link on the right hand side of the index, or combine edit/show somehow.
2. When loading the new host form with no hardware profiles in the DB, don't select a host group but then select a CR, you get a 404 error on the /hosts/compute_resource_selected AJAX call as it tries to find a host group where the ID is null.

Really cool, great work @isratrade!

domcleal commented on db/seeds.rb:
```diff
@@ -150,3 +150,17 @@ def audit_modified?(type, name)
+if HardwareProfile.unconfigured?
+  HardwareProfile.without_auditing do
+    [
+      { :name => '1-Small' },
```
I'm unsure about the value of this seed data, I'd probably prefer to leave it out as there's no corresponding CR-specific configuration.

ohadlevy replied:
the main reason i vote it in, is that it prefixed with an Int, which force
order.. I wanted to avoid handing defining the order of the hw profiles for
now as i think its non trivial?

domcleal replied:
Okay, then.

isratrade committed:
fixes #2317 - VM's with the name foreman_########

isratrade commented:
@domcleal, re-committed with following fixes
1.  gray 'unspecified' in hardware_profile#show
2.  fixed ajax error if new host form with no hardware profiles in the DB and don't select hostgroup
3.  moved menu to be above 'hardware models'
4.  "Show" in addition to "Edit" on index pages of hardware profiles
5.  Redirect to show page after creating new hardware profile.

isratrade commented:
- added table-two-pane with ease.  big thanks to @abenari +2
- added table-two-pane for #new as well with @ohadlevy's help
- wait on merging.  I want to add caching for pretty_vm_attrs, since it's very slow for the hardware_profiles#show page.

isratrade commented:
also added "Cancel" button on bottom of hardware_profiles#show page

domcleal commented:
When I click the "Show" button on the profiles index, it shows the edit page in two-pane mode?  While if I click on the name on the left, it doesn't do two-pane mode.

ohadlevy commented:
note that when switching to two-pane, you need to load all js correctly, atm i think most pages wont work correctly

ohadlevy commented:
@isratrade also note that audit messages seems odd, e.g.

```
 Admin (127.0.0.1) added Set (Hardware Profile / Compute Resource): sets/3

    Added N/A to N/A
```

ohadlevy commented:
also, caching for hw cr description at https://gist.github.com/8288942

isratrade commented:
@ohadlevy, @domcleal, new changes in 280e5fe to be squashed

isratrade commented:
@ohadlevy, I couldn't find an audit that was `Added N/A to N/A`
However, I did notice that it truncates the vm_attrs serialized hash

```
Item            Old                                                                  New
Vm attrs    {"flavor_id"=>"m1.large", "availability_zone"=>...  {"flavor_id"=>"m1.xlarge", "availability_zone"=...
```

isratrade commented:
notice the tab for #hardware_profiles in compute_resources#show

domcleal commented:
I see the N/As in audits when I did:
1. clean database
2. created a hardware profile "Test"
3. added a configuration on a libvirt CR

You can see the audits from 2) and 3) here:

![screenshot from 2014-01-07 10 39 32](https://f.cloud.github.com/assets/482089/1858230/46a8ee92-7788-11e3-89f1-bf4f3099a3f7.png)

Nice addition to compute resources!

isratrade committed:
fixes #3178 - API v2 - add compute profile to hosts and hostgroups response and create new VM without the need to manually pass [:host][:compute_attributes]

ohadlevy commented on app/models/host/managed.rb:
```diff
@@ -508,13 +508,18 @@ def jumpstart?
   end
 
+  def set_compute_attributes
+    return unless hardware_profile && compute_resource
```
1. should consider running it only if using settings[:attended] true
2. this would override the compute attributes provided by the form ? 
   maybe:

``` ruby
  return unless compute_attributes.empty?
  return unless hardware_profile_id && compute_resource_id
  self.compute_attributes = compute_resource.hardware_profile_defaults(hardware_profile_id)
```

isratrade replied:
ok, will update

isratrade committed:
fixes #3178 - add feature for compute profiles for hostgroups and hosts

isratrade commented:
recommitted per @ohadlevy's last comment.  All tests should pass.  I will submit separate PR's for API v2 hardware profiles and validation after this is merged.

isratrade commented on app/models/host/managed.rb:
```diff
@@ -139,9 +139,9 @@ class Jail < ::Safemode::Jail
                           :if => Proc.new { |host| host.managed and host.disk.empty? and not defined?(Rake) and capabilities.include?(:build) }
     validates :serial, :format => {:with => /[01],\d{3,}n\d/, :message => N_("should follow this format: 0,9600n8")},
                        :allow_blank => true, :allow_nil => true
+    after_validation :set_compute_attributes
```
this is now within `if SETTINGS[:unattended]` block

isratrade commented:
all tests passed!  ready for merge ... @domcleal, @ohadlevy

ohadlevy commented:
as discussed, @isratrade lets rename hardware_profile_compute_resource objects to compute_attributes, this hopefully should ease on the api usage.

ohadlevy commented:
and rename hardware_profile to compute_profile, as we save non hw specific attributes as well (like avail zone)

isratrade commented:
re-committed with name changes to files, code, and commit messages

ohadlevy commented on app/models/compute_attribute.rb:
```diff
@@ -0,0 +1,37 @@
+class ComputeAttribute < ActiveRecord::Base
+
+  include ActionView::Helpers::NumberHelper
```
do we need it here? I didnt see where we used it.

isratrade replied:
@ohadlevy, correct.  It was moved to FogExtensions and not used here.  I re-committed with fix.

isratrade commented:
@ohadlevy, @domcleal, tests passed after all the name changes :+1:

isratrade commented on app/services/menu/loader.rb:
```diff
@@ -64,6 +64,7 @@ def self.load
           menu.item :smart_proxies,       :caption => N_('Smart proxies')
           if SETTINGS[:unattended]
             menu.item :compute_resources, :caption => N_('Compute resources')
+            menu.item :compute_profiles,  :caption => N_('Compute profiles')
```
fyi, I put 'Compute profiles' under the 'Infrastructure' menu instead of Hosts > Provisioning Setup.

ohadlevy commented:
merged as c6e02bd thanks @isratrade!
